source/common/upstream/upstream_impl.cc

```diff
@@ -234,6 +232,9 @@ ClusterImplBase::ClusterImplBase(const envoy::api::v2::Cluster& cluster,
                                  Ssl::ContextManager& ssl_context_manager, bool added_via_api)
     : runtime_(runtime), info_(new ClusterInfoImpl(cluster, source_address, runtime, stats,
                                                    ssl_context_manager, added_via_api)) {
+  // Create the default (empty) priority set before registering callbacks to
+  // avoid getting an update the first time it is accessed.
```

Comment:
I actually had a question about this. I notice that in getOrCreate() you call runUpdateCallbacks() not matter what. Why is that the case? Could you add a comment there in a follow up?

Reply:
Hahaha, copy-paste from my follow-up HostSet& PrioritySetImpl::getOrCreateHostSet(uint32_t priority) {

      // TODO(alyssawilk) get Matt's input on if we want resize updates or no.
      // Right now one has to create a host set then add hosts.  Is it worth
      // notifying for the "no-op" add if there may not be hosts?
      runUpdateCallbacks(i, {}, {});


So right now we always runUpdateCallbacks() any time we change the priority set.  So if you getOrCreate(P=5) nd add hosts, you'll get updates for 2,3,4,5 [new hosts for 5]
My thought was it's worth noting when a host set is added but if there are no hosts, does it matter?  As long as all callback subscribers resize their data structures for 5 when they get the host update I think we can do away with the no-op callbacks.
Happy to make that change as a standalone or as part of the load balancer update.

Reply:
I see. IMO I would probably get rid of the NOP default callback unless we have a specific use case for it. This confused me when I read the code (though the explanation makes sense). I was actually worried about it in the context of the initial (0) set creation, and whether premature callbacks would do something bad during init. Again I could go either way on this as long as it is commented well, but my preference would be to go for simplicity.

Reply:
SG, will remove it in the follow-up and split it out if it's a larger change than I think it is.

No-op in the interim since no one should be using P!=0 and we don't have callbacks for P=0.

Reply:
We don't have callbacks for P=0 because the callbacks aren't installed yet right? (Not because there is a guard in the callback code)?

Reply:
Correct